backend/src/Parser/Parser.fs

```diff
@@ -504,49 +554,59 @@ let parseTestFile (filename : string) : Module =
       | "UUID" -> PT.TUuid
       | "Password" -> PT.TPassword
       | _ ->
-        Exception.raiseInternal
-          $"Unsupported type"
-          [ "name", ident.idText; "type", typ; "range", ident.idRange ]
+        match matchingUserTypes userTypes ident.idText with
+        | [ matched ] -> PT.TUserType matched
+        | _ ->
+          Exception.raiseInternal
+            $"Unsupported type"
+            [ "name", ident.idText; "type", typ; "range", ident.idRange ]
     | _ -> Exception.raiseInternal $"Unsupported type" [ "type", typ ]
 
-  let rec parseArgPat (pat : SynPat) : PT.UserFunction.Parameter =
+  let rec parseArgPat userTypes (pat : SynPat) : PT.UserFunction.Parameter =
     match pat with
-    | SynPat.Paren (pat, _) -> parseArgPat pat
+    | SynPat.Paren (pat, _) -> parseArgPat userTypes pat
     | SynPat.Const (SynConst.Unit, _) ->
       { id = gid (); name = "unit"; typ = PT.TUnit; description = "" }
     | SynPat.Typed (SynPat.Named (SynIdent (id, _), _, _, _), typ, _) ->
-      { id = gid (); name = id.idText; typ = convertType typ; description = "" }
+      { id = gid ()
+        name = id.idText
+        typ = convertType userTypes typ
+        description = "" }
     | _ -> Exception.raiseInternal "Unsupported argPat" [ "pat", pat ]
 
-  let parseSignature (pat : SynPat) : string * List<PT.UserFunction.Parameter> =
+  let parseSignature
+    userTypes
+    (pat : SynPat)
+    : string * List<PT.UserFunction.Parameter> =
     match pat with
     | SynPat.LongIdent (SynLongIdent ([ name ], _, _), _, _, argPats, _, _) ->
       let parameters =
         match argPats with
-        | SynArgPats.Pats pats -> List.map parseArgPat pats
+        | SynArgPats.Pats pats -> List.map (parseArgPat userTypes) pats
         | SynArgPats.NamePatPairs _ ->
           Exception.raiseInternal "Unsupported pattern" [ "pat", pat ]
       (name.idText, parameters)
     | _ -> Exception.raiseInternal "Unsupported pattern" [ "pat", pat ]
 
-  let parseBinding (binding : SynBinding) : PT.UserFunction.T =
+  let parseBinding userTypes (binding : SynBinding) : PT.UserFunction.T =
```

Comment:
Similarly, I'm pretty sure we shouldn't be passing userTypes into the parser (like we don't pass userFunctions into the parser)

Reply:
Without context of the types, how is the parser supposed to figure out if something is an unnamed variable or an enum case?

Reply:
when parsed as F#, it doesn't present it to our 'converter' as "this is an enum case" - it's just an identifier.  Passing in UserTypes allows us to determine whether it should be an EConstructor or an EVariable. I'm not sure how else we could handle this, but maybe I'm missing something.

Reply:
I see - that's a tricky problem. I guess there's a similar problem for function names? Or can they tell from context?

Reply:
Yes, there's a related situation with fns right now - `DateTime.now` is parsed as a LongIdent in F# AST, and we currently assume all LongIdents are fn calls. If/when we want to refer to a fn within a specific module, we'll face the same situation.

Reply:
clarification on the above comment:

> we currently assume all LongIdents are fn calls

is not fully true.

This is a better phrasing:
Given a LongIdent:
- if it has 2 components (A.B) and the first character of the first component is uppercase, then it's a fn call\
- (kinda supported: if it has 4 components (A.B.C.D), then it's a package fn call)
- otherwise, it's some sort of field access

Reply:
I guess in F# they're all values, and the typechecker really is the thing that checks whether it's valid to apply it

Reply:
I suppose I'd prefer if we had a separate step after the parser to resolve what's a type and what's a function, but I'm not sure that's possible?

Reply:
I'm not sure it's possible either. Maybe it's worth following up on that later? I'm sort of stumped of how we could do this another way.